File: week-1/problem-set-recursion/prob7.py
"""Prob7 example.

This file belongs to the week 1 recursion practice solutions. It uses replace to illustrate the main idea with a small runnable example.
"""

def replace(s, from_char, to_char):
    """This function recursively replaces all occurrences of a character in a string with another character.

    Args:
        s (str): The s input used by this function.
        from_char (str): The from_char input used by this function.
        to_char (str): The to_char input used by this function.

    Returns:
        str: The string produced by the algorithm.
    """
    if not s:
        # Base case: the string is empty
        return s
    
    if s[0] == from_char:
        # Recursive case: replace the first character if it matches the from_char
        return to_char + replace(s[1:], from_char, to_char)
    else:
        # Recursive case: keep the first character and continue with the rest of the string
        return s[0] + replace(s[1:], from_char, to_char)

s = "steve"
from_char = 'e'
to_char = 'a'

s = replace(s, from_char, to_char)

print(s)  # Output: "stava"

File: week-1/problem-set-recursion/prob7.cpp
/**
 * @file prob7.cpp
 * @brief C++ version of the recursive character-replacement example.
 */

#include <iostream>
#include <string>

using namespace std;

/**
 * @brief Recursively replace every occurrence of one character with another.
 * @param text const string& Original text that may contain the character to replace.
 * @param from_char char Character that should be replaced.
 * @param to_char char Replacement character.
 * @return string New string after every matching character has been replaced.
 */
string replace_recursive(const string& text, char from_char, char to_char) {
    if (text.empty()) {
        return text;
    }

    if (text[0] == from_char) {
        return string(1, to_char) + replace_recursive(text.substr(1), from_char, to_char);
    }
    return string(1, text[0]) + replace_recursive(text.substr(1), from_char, to_char);
}

/**
 * @brief Run the recursive replacement demonstration.
 * @return int Exit status code returned to the operating system.
 */
int main() {
    string text = "steve";
    char from_char = 'e';
    char to_char = 'a';

    cout << replace_recursive(text, from_char, to_char) << '\n';
    return 0;
}
